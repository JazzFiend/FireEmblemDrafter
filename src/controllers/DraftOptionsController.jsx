import React, { Component } from 'react';
import PropTypes from 'prop-types';

import DraftOptions from '../view/DraftOptions';

class DraftOptionsController extends Component {
  constructor(props) {
    super(props);
    this.state = {
      errorMessage: '',
    };
  }

  handleTeamSizeChange(teamSize) {
    const { handleTeamSizeChange, maxTeamSize } = this.props;
    if (teamSize <= 0 || teamSize > maxTeamSize) {
      this.setState({
        errorMessage: 'Please enter a valid team size.',
      });
    } else {
      this.setState({
        errorMessage: '',
      });
    }
    handleTeamSizeChange(teamSize);
  }

  showDraftOptions() {
    const { gameSelected, draftInProgress } = this.props;
    return !draftInProgress && gameSelected;
  }

  render() {
    const { defaultValue } = this.props;
    const { errorMessage } = this.state;
    return (
      <DraftOptions
        showDraftOptions={this.showDraftOptions()}
        handleTeamSizeChange={(teamSize) => this.handleTeamSizeChange(teamSize)}
        defaultValue={defaultValue}
        errorMessage={errorMessage}
      />
    );
  }
}
export default DraftOptionsController;

DraftOptionsController.propTypes = {
  gameSelected: PropTypes.bool.isRequired,
  draftInProgress: PropTypes.bool.isRequired,
  defaultValue: PropTypes.number,
  maxTeamSize: PropTypes.number.isRequired,
  handleTeamSizeChange: PropTypes.func,
};

DraftOptionsController.defaultProps = {
  defaultValue: 0,
  handleTeamSizeChange: () => {},
};
