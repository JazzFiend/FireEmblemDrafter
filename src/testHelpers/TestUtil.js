import configureStore from 'redux-mock-store';

export default class TestUtil {
  static createDraftInProgressMockStore(inProgress) {
    const mockStore = configureStore([]);
    return mockStore({
      draftInProgress: {
        value: inProgress,
      },
    });
  }
}
