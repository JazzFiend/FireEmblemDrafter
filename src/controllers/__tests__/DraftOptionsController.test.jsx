import {
  render,
  cleanup,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { Provider } from 'react-redux';
import DraftOptionsController from '../DraftOptionsController';
import TestUtil from '../../testHelpers/TestUtil';

afterEach(cleanup);

test('does not render draft options when game has not been selected', () => {
  const store = TestUtil.createDraftInProgressMockStore(false);
  const { asFragment } = render(
    <Provider store={store}>
      <DraftOptionsController
        gameSelected={false}
        defaultValue={10}
        maxTeamSize={99}
      />
    </Provider>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('does not render draft options when draft is in progress', () => {
  const store = TestUtil.createDraftInProgressMockStore(true);
  const { asFragment } = render(
    <Provider store={store}>
      <DraftOptionsController gameSelected={true} defaultValue={20} maxTeamSize={99} />
    </Provider>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders draft options when game is selected and draft is not in progress', () => {
  const store = TestUtil.createDraftInProgressMockStore(false);
  const { asFragment } = render(
    <Provider store={store}>
      <DraftOptionsController gameSelected={true} defaultValue={30} maxTeamSize={99} />
    </Provider>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('saves state of team size on change', () => {
  const dummy = jest.fn();
  const store = TestUtil.createDraftInProgressMockStore(false);
  const { getByTestId } = render(
    <Provider store={store}>
      <DraftOptionsController
        gameSelected={true}
        handleTeamSizeChange={dummy}
        defaultValue={0}
        maxTeamSize={99}
      />
    </Provider>,
  );
  userEvent.type(getByTestId('team-size-textbox'), '3');

  expect(dummy).toHaveBeenCalledWith(3);
});

test('displays error when team size is invalid', () => {
  const dummy = jest.fn();
  const store = TestUtil.createDraftInProgressMockStore(false);
  const { asFragment, getByTestId } = render(
    <Provider store={store}>
      <DraftOptionsController
        gameSelected={true}
        handleTeamSizeChange={dummy}
        defaultValue={0}
        maxTeamSize={99}
      />
    </Provider>,
  );
  userEvent.type(getByTestId('team-size-textbox'), '0');

  expect(asFragment()).toMatchSnapshot();
});

test('displays error when team size is too big', () => {
  const dummy = jest.fn();
  const store = TestUtil.createDraftInProgressMockStore(false);
  const { asFragment, getByTestId } = render(
    <Provider store={store}>
      <DraftOptionsController
        gameSelected={true}
        handleTeamSizeChange={dummy}
        defaultValue={10}
        maxTeamSize={10}
      />
    </Provider>,
  );
  userEvent.type(getByTestId('team-size-textbox'), '10');

  expect(asFragment()).toMatchSnapshot();
});
