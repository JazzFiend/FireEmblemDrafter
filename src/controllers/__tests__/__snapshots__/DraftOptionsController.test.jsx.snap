// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays error when team size is invalid 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        Team Size:
        <input
          data-testid="team-size-textbox"
          name="teamSize"
          type="number"
          value="0"
        />
      </form>
      <div
        class="error"
        style="color: red;"
      >
        Please enter a valid team size.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`displays error when team size is too big 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        Team Size:
        <input
          data-testid="team-size-textbox"
          name="teamSize"
          type="number"
          value="10"
        />
      </form>
      <div
        class="error"
        style="color: red;"
      >
        Please enter a valid team size.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`does not render draft options when draft is in progress 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`does not render draft options when game has not been selected 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`renders draft options when game is selected and draft is not in progress 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        Team Size:
        <input
          data-testid="team-size-textbox"
          name="teamSize"
          type="number"
          value="30"
        />
      </form>
      <div
        class="error"
        style="color: red;"
      />
    </div>
  </div>
</DocumentFragment>
`;
